File: cloudbank/proxy/src/proxy/protocol.py
"""Browser ↔ proxy WebSocket message types.

Mirrors cloudbank/web/src/agent/protocol.ts. Any change here must land there
too — the contract is symmetric.
"""
from __future__ import annotations

import json
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field, TypeAdapter, ValidationError

Persona = Literal["chloe", "david", "tom", "sarah"]

# Cap on a single inbound browser frame. Audio arrives as small base64 PCM16
# chunks (a 100ms 16 kHz frame is ~4 KB), so 1 MiB is generous. Callers check
# this before parsing; the field limits below bound the parsed payloads too.
MAX_CLIENT_FRAME_BYTES = 1024 * 1024
_MAX_AUDIO_B64_LEN = MAX_CLIENT_FRAME_BYTES
_MAX_TEXT_LEN = 8192


# Client → proxy
class ClientStart(BaseModel):
    type: Literal["start"]
    persona: Persona
    # Session variables forwarded to CES as setSessionState entries.
    # The client (web) computes them from the persona fixture and sends
    # them on every start frame — the proxy is a dumb relay.
    variables: dict[str, str]


class AudioFrame(BaseModel):
    """Used in both directions; type='audio'."""
    type: Literal["audio"]
    data: str = Field(max_length=_MAX_AUDIO_B64_LEN)  # base64 PCM16


class ClientText(BaseModel):
    type: Literal["text"]
    text: str = Field(max_length=_MAX_TEXT_LEN)


class ClientToolResponse(BaseModel):
    type: Literal["tool_response"]
    id: str
    response: dict[str, Any]  # {"output": {...}} or {"error": "..."}


class ClientStop(BaseModel):
    type: Literal["stop"]


ClientMsg = Annotated[
    ClientStart | AudioFrame | ClientText | ClientToolResponse | ClientStop,
    Field(discriminator="type"),
]

_client_adapter: TypeAdapter[ClientMsg] = TypeAdapter(ClientMsg)


_UNKNOWN_TYPE_CODES = {"union_tag_invalid", "union_tag_not_found"}


def parse_client_msg(raw: str) -> ClientMsg:
    try:
        return _client_adapter.validate_json(raw)
    except ValidationError as exc:
        # Surface a friendly error when the `type` discriminator is unknown,
        # so callers don't have to inspect Pydantic's internal error model.
        for err in exc.errors():
            if err["type"] in _UNKNOWN_TYPE_CODES:
                raise ValueError("unknown client message") from exc
        raise ValueError(str(exc)) from exc


# Proxy → client
class ServerReady(BaseModel):
    type: Literal["ready"] = "ready"


class ServerTranscript(BaseModel):
    type: Literal["transcript"] = "transcript"
    text: str
    isFinal: bool = False


class ServerText(BaseModel):
    type: Literal["text"] = "text"
    text: str


class ServerClientFunction(BaseModel):
    type: Literal["client_function"] = "client_function"
    id: str
    name: str
    args: dict[str, Any]


class ServerInterrupt(BaseModel):
    type: Literal["interrupt"] = "interrupt"


class ServerEnd(BaseModel):
    type: Literal["end"] = "end"


class ServerError(BaseModel):
    type: Literal["error"] = "error"
    message: str


ServerMsg = (
    ServerReady | ServerTranscript | ServerText | AudioFrame
    | ServerClientFunction | ServerInterrupt | ServerEnd | ServerError
)


def serialize_server_msg(msg: ServerMsg) -> str:
    return json.dumps(msg.model_dump(mode="json"), separators=(",", ":"))

File: cloudbank/web/src/agent/protocol.ts
// Browser ↔ proxy WS protocol. Mirrors cloudbank/proxy/src/proxy/protocol.py.

export type ClientMsg =
  // `sessionId` names the CES conversation to (re)join. The browser owns it so
  // that a reconnect after CES's ~20-30s inactivity close resumes the same
  // conversation rather than opening a blank one. See agent/sessionId.ts.
  | { type: 'start'; persona: 'chloe' | 'david' | 'tom' | 'sarah'; variables: Record<string, string>; sessionId: string }
  | { type: 'audio'; data: string } // base64 PCM16 16k
  | { type: 'text'; text: string }
  | { type: 'tool_response'; id: string; response: { output: Record<string, unknown> } | { error: string } }
  | { type: 'stop' }

export type ServerMsg =
  | { type: 'ready' }
  | { type: 'transcript'; text: string; isFinal: boolean }
  | { type: 'text'; text: string }
  | { type: 'audio'; data: string } // base64 PCM16 24k
  | { type: 'client_function'; id: string; name: string; args: Record<string, unknown> }
  | { type: 'interrupt' }
  | { type: 'end' }
  | { type: 'error'; message: string }

const SERVER_TYPES = new Set([
  'ready', 'transcript', 'text', 'audio', 'client_function', 'interrupt', 'end', 'error',
])

export function parseServerMsg(raw: string): ServerMsg {
  const parsed = JSON.parse(raw)
  if (!parsed || typeof parsed !== 'object' || typeof parsed.type !== 'string') {
    throw new Error('unknown server message: missing type')
  }
  if (!SERVER_TYPES.has(parsed.type)) {
    throw new Error(`unknown server message: ${parsed.type}`)
  }
  // Validate the payload of the variants whose fields are dereferenced without
  // a guard downstream. Checking the discriminator alone is not enough: a
  // {"type":"audio"} frame with no `data` would reach atob(undefined) inside an
  // async onmessage handler, surfacing as an unhandled rejection rather than a
  // caught, logged parse failure.
  switch (parsed.type) {
    case 'audio':
      if (typeof parsed.data !== 'string') throw new Error('audio frame: missing data')
      break
    case 'transcript':
    case 'text':
      if (typeof parsed.text !== 'string') throw new Error(`${parsed.type} frame: missing text`)
      break
    case 'error':
      if (typeof parsed.message !== 'string') throw new Error('error frame: missing message')
      break
    case 'client_function':
      if (typeof parsed.id !== 'string' || typeof parsed.name !== 'string') {
        throw new Error('client_function frame: missing id/name')
      }
      if (parsed.args !== undefined && (typeof parsed.args !== 'object' || parsed.args === null)) {
        throw new Error('client_function frame: args must be an object')
      }
      break
  }
  return parsed as ServerMsg
}

export function serializeClientMsg(msg: ClientMsg): string {
  return JSON.stringify(msg)
}
